$NetBSD: patch-al,v 1.2 2003/02/15 16:29:33 jmmv Exp $

This patch is being tracked in bug #92844.
See http://bugzilla.gnome.org/show_bug.cgi?id=92844 for more details.

--- modules/file-method.c.orig	2003-02-12 12:55:48.000000000 +0100
+++ modules/file-method.c
@@ -488,10 +488,10 @@ do_truncate (GnomeVFSMethod *method,
 
 typedef struct {
 	GnomeVFSURI *uri;
-	DIR *dir;
+	GDir *dir;
 	GnomeVFSFileInfoOptions options;
 
-	struct dirent *current_entry;
+	gchar *current_entry;
 
 	gchar *name_buffer;
 	gchar *name_ptr;
@@ -499,7 +499,7 @@ typedef struct {
 
 static DirectoryHandle *
 directory_handle_new (GnomeVFSURI *uri,
-		      DIR *dir,
+		      GDir *dir,
 		      GnomeVFSFileInfoOptions options)
 {
 	DirectoryHandle *result;
@@ -511,9 +511,6 @@ directory_handle_new (GnomeVFSURI *uri,
 	result->uri = gnome_vfs_uri_ref (uri);
 	result->dir = dir;
 
-	/* Reserve extra space for readdir_r, see man page */
-	result->current_entry = g_malloc (sizeof (struct dirent) + GET_PATH_MAX() + 1);
-
 	full_name = get_path_from_uri (uri);
 	g_assert (full_name != NULL); /* already done by caller */
 	full_name_len = strlen (full_name);
@@ -538,7 +535,6 @@ directory_handle_destroy (DirectoryHandl
 {
 	gnome_vfs_uri_unref (directory_handle->uri);
 	g_free (directory_handle->name_buffer);
-	g_free (directory_handle->current_entry);
 	g_free (directory_handle);
 }
 
@@ -737,13 +733,13 @@ do_open_directory (GnomeVFSMethod *metho
 		   GnomeVFSContext *context)
 {
 	gchar *directory_name;
-	DIR *dir;
+	GDir *dir;
 
 	directory_name = get_path_from_uri (uri);
 	if (directory_name == NULL)
 		return GNOME_VFS_ERROR_INVALID_URI;
 
-	dir = opendir (directory_name);
+	dir = g_dir_open (directory_name, 0, NULL);
 	g_free (directory_name);
 	if (dir == NULL)
 		return gnome_vfs_result_from_errno ();
@@ -764,7 +760,7 @@ do_close_directory (GnomeVFSMethod *meth
 
 	directory_handle = (DirectoryHandle *) method_handle;
 
-	closedir (directory_handle->dir);
+	g_dir_close (directory_handle->dir);
 
 	directory_handle_destroy (directory_handle);
 
@@ -784,20 +780,9 @@ do_read_directory (GnomeVFSMethod *metho
 
 	handle = (DirectoryHandle *) method_handle;
 	
-	errno = 0;
-	if (readdir_r (handle->dir, handle->current_entry, &result) != 0) {
-		/* Work around a Solaris bug.
-		 * readdir64_r returns -1 instead of 0 at EOF.
-		 */
-		if (errno == 0) {
-			return GNOME_VFS_ERROR_EOF;
-		}
-		return gnome_vfs_result_from_errno ();
-	}
-	
-	if (result == NULL) {
+	handle->current_entry = g_dir_read_name (handle->dir);
+	if (handle->current_entry == NULL)
 		return GNOME_VFS_ERROR_EOF;
-	}
 
 	file_info->name = g_strdup (result->d_name);
 
